postproengine/__init__.py

```diff
@@ -199,6 +200,65 @@ def print_inputs(subset=[], plist=pluginlist):
         print()
     return
 
+# =====================================================
+# --- ADD REUSABLE CLASSES AND METHODS HERE ---
+# =====================================================
+
+def compute_axis1axis2_coords(db):
+    """
+    Computes the native axis1 and axis2 coordinate system for a given
+    set of sample planes.
+
+    Note: this assumes db has the origin, axis1/2, and offset definitions
+    """
+
+    # Check to make sure db has everything needed
+    if ('origin' not in db) or \
+       ('axis1' not in db) or \
+       ('axis2' not in db) or \
+       ('axis3' not in db) or \
+       ('offsets') not in db:
+        print('Need to ensure that the sample plane data includes origin, axis1, axis2, axis3, and offset information')
+        return
+
+    # Pull out the coordate definitions
+    axis1  = np.array(db['axis1'])
+    axis2  = np.array(db['axis2'])
+    axis3  = np.array(db['axis3'])
+    origin = np.array(db['origin'])
+    offsets= db['offsets']
+    if not isinstance(offsets,list):
+        offsets = [offsets]
+    
+    # Compute the normals
+    n1 = axis1/np.linalg.norm(axis1)
+    n2 = axis2/np.linalg.norm(axis2)
+    if np.linalg.norm(axis3) > 0.0:
+        n3 = axis3/np.linalg.norm(axis3)
+    else:
+        n3 = axis3
+        
```

Comment:
Yes, we can definitely do it that way as long as we don't need anything that requires `R` to be inverted (the `n3` vector might be 0).  I used the slow, dumb way to compute a1, a2, a3 because I didn't trust myself to code the offset subtraction for each plane in the proper pythonic way, but we try it and test it out against this known method.

I might do the u_a1/u_a2/u_a3 conversion in a separate routine because that part of it could be generalized to handle any velocity-like structures (e.g., instantaneous velocities, mean velocities, etc.), and you may not always need coordinate transformation and velocity transformations at the same time.

Lawrence

Reply:
Are (the normalized) axis1, axis2, and axis3 guaranteed to be orthonormal? If so, can we just compute the curl of axis1 and axis2 to get axis3 and avoid discrepancies in the way the offset-direction is handled in the pre-processing step? Then R should almost always just be a rotation of the Cartesian x,y and z, which is invertible if that's needed. 

Separating the velocity conversion makes sense to me! We should form the mapping once and then apply it as needed.

Reply:
Unfortunately, `axis1`, `axis2`, and `axis3` are not guaranteed to be orthogonal.  We did this on purpose because the user might want `axis3` to follow a separate direction compared to the plane normal (i.e., the sample planes could be rotor planes, but because of yaw, axis3 might be different compared to the sample plane normal).  Also for the case of just a single 2D plane, axis3 ends up just defaulting to (0,0,0).

Reply:
Okay, got it, thanks for clarifying. I don't think we'll need to invert R so using n1,n2 and n3 should be sufficient.

Reply:
Generalized this so that any list of arbitrary points can be converted in `convert_pt_xyz_to_axis1axis2()` using matrix multiplication routines.